### ISSUE TITLE
Documentation says BIO_s_dgram_pair is thread-safe, but BIO relies on single-threaded patterns

### ISSUE THREAD

**Issue by davidben** [openssl/openssl#26059]
I noticed the docs for `BIO_s_dgram_pair` say:

> [BIO_recvmmsg(3)](https://docs.openssl.org/master/man3/BIO_sendmmsg/), [BIO_sendmmsg(3)](https://docs.openssl.org/master/man3/BIO_sendmmsg/), [BIO_read(3)](https://docs.openssl.org/master/man3/BIO_read/), [BIO_write(3)](https://docs.openssl.org/master/man3/BIO_read/), [BIO_pending(3)](https://docs.openssl.org/master/man3/BIO_ctrl/), [BIO_get_write_guarantee(3)](https://docs.openssl.org/master/man3/BIO_s_bio/) and [BIO_flush(3)](https://docs.openssl.org/master/man3/BIO_ctrl/) may be used by multiple threads simultaneously on the same BIO datagram pair. Specific [BIO_ctrl(3)](https://docs.openssl.org/master/man3/BIO_ctrl/) operations (namely BIO_CTRL_PENDING, BIO_CTRL_FLUSH and BIO_C_GET_WRITE_GUARANTEE) may also be used. Invoking any other BIO call, or any other [BIO_ctrl(3)](https://docs.openssl.org/master/man3/BIO_ctrl/) operation, on either half of a BIO datagram pair while any other BIO call is also in progress to either half of the same BIO datagram pair results in undefined behaviour.

This was surprising to me given how `BIO` works. I'm not sure this is actually meaningfully thread-safe. Although the `BIO` does internally locked, we have a couple of single-threaded patterns in `BIO`:

First, the `BIO_read` and `BIO_write` wrappers, without any locks, update some state: https://github.com/openssl/openssl/blob/master/crypto/bio/bio_lib.c#L288 and https://github.com/openssl/openssl/blob/master/crypto/bio/bio_lib.c#L359

More fundamentally, the `BIO` API itself is single-threaded. Instead of returning everything directly, `BIO` likes to store outputs on the `BIO` for the caller to query later, notably [retry flags](https://github.com/openssl/openssl/blob/master/crypto/bio/bss_dgram_pair.c#L1210). The new `BIO_read_ex` APIs seem to have gone further in this direction in incompatibly replacing an in-band EOF return with out-of-band EOF state. See https://github.com/openssl/openssl/issues/8208.

This is a single-threaded API pattern: even if language-level race conditions are eliminated with locking, the application still has a race condition if another thread uses the `BIO` in between, say, `BIO_read` and `BIO_should_read`.

**Comment by t8m** [comment]
Yeah, I think this is a mistake, actually only  the [BIO_recvmmsg(3)](https://docs.openssl.org/master/man3/BIO_sendmmsg/) and [BIO_sendmmsg(3)](https://docs.openssl.org/master/man3/BIO_sendmmsg/) are supposed to be simultaneously usable.

**Comment by andrewkdinh** [comment]
@davidben are you able to open a PR to update the documentation for this? Would be greatly appreciated

**Comment by davidben** [comment]
That's probably better done by someone from the OpenSSL project. I don't have the context behind `BIO_s_dgram_pair`. I can flag when, based on my familiarity with a different offshoot of this codebase, something in OpenSSL seems wrong. But what exactly would be right (e.g. which of those operations are expected to be thread-safe), I'm not best to say. I can simply advise that your documentation doesn't make sense for API with a connection to `BIO_should_*`.
